### PR TITLE
[Perf] Add TRTLLM FP8 MoE Modular Kernel

### PR DESCRIPTION
**Pull request by wzhao18**
## Purpose
Add TRTLLM FP8 MoE Modular Kernel. The trtllm monolithic kernel has restrictions over the routing method of the model. This PR enables more models (e.g., minimax m2) to use the trtllm moe backend by adding the modular version.

## Test Plan
Tested with Minimax-m2.5 (which cannot use trtllm moe monolithic backend due to routing method restriction)
```
vllm serve MiniMaxAI/MiniMax-M2.5 \
    --trust-remote-code \
    --tensor-parallel-size 2 \
    --enable-expert-parallel
```

Note: to get Minimax to run with trtllm moe, it is required to set minimax's [router_logits_dtype](https://github.com/vllm-project/vllm/blob/755356b3d18d8079b1b115dfd2111dc1accdb764/vllm/model_executor/models/minimax_m2.py#L110) to `bfloat16`, as trtllm backend only supports this. This restriction should go away soon with [flashinfer PR](https://github.com/flashinfer-ai/flashinfer/pull/2534). (Edit 03/17: this is not true. the modular kernel does not enforce router logits dtype, as routing is done externally. The constraint should be removed)

## Test Result
1K/1K TP=2 Benchmark on B200:
isl1024_osl1024

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py [hunk 1]
```diff
@@ -50,6 +55,170 @@ def __init__(
...
+            gemm2_weights_scale=self.quant_config.w2_scale,
+            num_experts=global_num_experts,
+            top_k=self.topk,
+            n_group=None,
+            topk_group=None,
+            intermediate_size=self.intermediate_size_per_partition,
+            local_expert_offset=self.ep_rank * self.local_num_experts,
+            local_num_experts=self.local_num_experts,
+            routed_scaling_factor=None,
+            routing_method_type=1,
```
[the remark this answers is not in the document]
Comment (wzhao18) [reply vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py:199]: `routing_method_type` is not actually used by the kernel in the routed case. Setting to 1 will work. This is done in the [nvfp4 case](https://github.com/wzhao18/vllm/blob/c012a8c477dd78b4444f22568b2bf1b08f2ad813/vllm/model_executor/layers/fused_moe/experts/trtllm_nvfp4_moe.py#L205) as well.

File: vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py [hunk 2]
```diff
@@ -50,6 +55,170 @@ def __init__(
...
+            return
+
+        # `trtllm_fp8_block_scale_routed_moe` has a bug and does not write to the
+        # output tensor in-place so we need to manually copy the result to the
+        # output tensor.
+        result = flashinfer.fused_moe.trtllm_fp8_block_scale_routed_moe(
+            topk_ids=packed_topk_ids,
+            routing_bias=None,
+            hidden_states=hidden_states,
+            hidden_states_scale=a1q_scale.t().contiguous(),  # type: ignore[union-attr]
```
[the remark this answers is not in the document]
Comment (wzhao18) [reply vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py:186]: fixed.

File: vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py [hunk 3]
```diff
@@ -44,6 +49,173 @@ def __init__(
...
+
+        # `trtllm_fp8_block_scale_routed_moe` has a bug and does not write to the
+        # output tensor in-place so we need to manually copy the result to the
+        # output tensor
+        # https://github.com/flashinfer-ai/flashinfer/issues/2703
+        result = flashinfer.fused_moe.trtllm_fp8_block_scale_routed_moe(
+            topk_ids=packed_topk_ids,
+            routing_bias=None,
+            hidden_states=hidden_states,
+            hidden_states_scale=a1q_scale.t().contiguous(),  # type: ignore[union-attr]
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py:186]: Could we fuse this from the activation quant before? This could be quite slow. Worth considering when fixing the output issue

File: vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py [hunk 4]
```diff
@@ -44,6 +49,173 @@ def __init__(
...
+            gemm2_weights_scale=self.quant_config.w2_scale,
+            num_experts=global_num_experts,
+            top_k=self.topk,
+            n_group=None,
+            topk_group=None,
+            intermediate_size=self.intermediate_size_per_partition,
+            local_expert_offset=self.ep_rank * self.local_num_experts,
+            local_num_experts=self.local_num_experts,
+            routed_scaling_factor=None,
+            routing_method_type=1,
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py:199]: Can you note that this is ignored in this case?

**Review by mgoin — APPROVED** [review]
LGTM as an interim step then, thanks!

File: vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py (hunk 3 above)
Comment (wzhao18) [reply vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py:186]: Will keep a note about it. Thanks for point this out.

File: vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py (hunk 4 above)
Comment (wzhao18) [reply vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py:199]: Will do in future PR then.
